ETag for static file server

Spixmaster opened the pull request:
Programmed feature https://github.com/yhirose/cpp-httplib/issues/2242. First off, the static file server now sends HTTP response header "ETag". Following HTTP requests by the client which include HTTP request header "If-None-Match" are only served if the value for HTTP response header "ETag" is not included in the value of HTTP request header "If-None-Match", otherwise an HTTP response with status code 304 is served which includes the HTTP response header "ETag" again that would have been sent with a normal status code of 200.

Useful resources:
- https://http.dev/caching
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/ETag
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/If-None-Match
- https://www.rfc-editor.org/rfc/rfc9110.html#name-304-not-modified

Spixmaster commented:
```sh
100% tests passed, 0 tests failed out of 444

Total Test time (real) = 365.31 sec

The following tests did not run:
	114 - BindServerTest.BindDualStack (Disabled)
	390 - SSLClientServerTest.LargeDataTransfer (Disabled)
```

Spixmaster commented:
The solution was done with one line. A c-string did not have null termination.

```sh
100% tests passed, 0 tests failed out of 444

Total Test time (real) = 365.31 sec

The following tests did not run:
	114 - BindServerTest.BindDualStack (Disabled)
	390 - SSLClientServerTest.LargeDataTransfer (Disabled)
```

Xxproner reviewed (commented):
lgfm. But I suggest you to move etag generation algo from `handle_file_request` and add func like `add_etag_enable()` what set the algo. class Server (its user) should know about the extension. Explicit is better than implicit.

And second moment. On `7924` line you check `etag.length() >= 2 ` and use then `.at(0)` and `.at(1)` that check bounds: `.length() > 0` and `.length() > 1`, it is useless :) (ref: https://en.cppreference.com/w/cpp/string/basic_string/at.html).

These remarks do not affect logic, it is correct. And your work is great.

Spixmaster commented:
Thank you for the review. I am glad that this feature is well perceived.

Regarding your first recommendation, do you suggest a switch (member variable on class `Server`) which serves as a condition to excute the HTTP response header `ETag`? Or do you suggest to have a member variable of type `std::optional<std::function<std::string(std::string)>` on class `Server` and additionally the member function `void add_etag_enable(const std::function<std::string(std::string)> f)` which assigns an value to the member variable which is then used to generated the `ETag` HTTP response header? In case of `std::nullopt`, the generation would be disabled.

From my current knowledge, `ETag` should always be sent to save bandwidth when interacting with browsers. It is also on by default in Nginx, https://nginx.org/en/docs/http/ngx_http_core_module.html#etag. Since we do not use the in my opinion inferior `If-Modified-Since` alternatively, I see no reason to allow disabling it.

One reason for disabling might be due to performance.

Do I even understand you correctly? What do you think?

Regarding your second recommendation, I know that `std::string::at()` has bound checks and would throw exception. However, no exception should be thrown. A client could send `If-None-Match: a` and the the file serving would throw although the HTTP response is valid.

I think the condition needs to be like before `if (etag.length() >= 2 && etag.at(0) == 'W' && etag.at(1) == '/')`.

Normally, I would use `std::string::starts_with()` but C++20 or newer is not used.

Spixmaster commented:
Done. Please review. All tests pass either with or without macro `CPPHTTPLIB_OPENSSL_SUPPORT`.

While the implementation is fine, please take a look at minor things I just assumed and guessed.

- name and location of member variable `is_etag_enabled`
- name and location of member function (declaration and definition) `get_is_etag_enabled()`
- name and location of member function (declaration and definition) `set_is_etag_enabled()`
- name and location of `TEST(StaticFileSever, If_Match)`
- name and location of `TEST(StaticFileSever, If_None_Match)`

Adjust to have consistency. These are things maintainers need to do.

Preferably, add your own commits.

Spixmaster commented:
The result from Claude is good.

Please adjust for consistency. I had to make `st_` from `httplib::detail::FileStat` public.

### Discovered Issues: issue 1

Proper issue. Fixed it.

### Serious Performance Issues: issue 1

 I did not verify but it is logical. I changed the ETag constitution to the one from Nginx which I verified with https://stackoverflow.com/questions/40504281/c-how-to-check-the-last-modified-time-of-a-file and https://serverfault.com/questions/690341/algorithm-behind-nginx-etag-generation.

### Serious Performance Issues: issue 2

Just listing existing constitutions. No real issue.

### Serious Performance Issues: issue 3

Fixed with second commit, https://github.com/yhirose/cpp-httplib/pull/2243/commits/dc30799710a8a2d55397cd43c105896029f8c84c. Although I do not see, that whitspace is allowed in https://www.rfc-editor.org/rfc/rfc9110.html#name-etag.

### Serious Performance Issues: issue 4

Not solvable and not a huge issue.

Spixmaster commented:
In summary, take a special look at the following and adjust for consistency.

- name and location of member variable `is_etag_enabled`
- name and location of member function (declaration and definition) `get_is_etag_enabled()`
- name and location of member function (declaration and definition) `set_is_etag_enabled()`
- name and location of `TEST(StaticFileSever, If_Match)`
- name and location of `TEST(StaticFileSever, If_None_Match)`
- way of disabling tests; currently macros are used as condition
- access specifier of `st_` and `ret_`; maybe, add another function (getter) that returns `st_` but if `ret_ < 0` returns `nullptr`

Spixmaster commented:
What about the current state of approval for this?

yhirose commented:
Copilot still doesn't like this code change...

# Code Quality Review for PR #2243

## 🔴 Critical Issues

### 1. Inappropriate Use of Exceptions
```cpp
inline std::uint64_t FileStat::last_modified() const {
  if (is_dir() || is_file()) {
    // ... 
  } else {
    throw std::runtime_error("Invalid directory or file.");
  }
}
```
- This library generally avoids exceptions by design, yet `std::runtime_error` is thrown here
- Returning `std::optional<uint64_t>` or returning `0` on error would be more consistent with the library's conventions

### 2. ETag Calculation Vulnerability
```cpp
const std::string etag = R"(")" + detail::from_i_to_hex(stat.last_modified()) + "-" +
                         detail::from_i_to_hex(mm->size()) + R"(")";
```
- If file content changes but size and modification time remain the same, the ETag will be identical
- Using a content hash (MD5/SHA1, etc.) would be more reliable

---

## 🟡 Moderate Issues

### 3. Verbose Method Naming
```cpp
bool get_is_etag_enabled() const;
Server &set_is_etag_enabled(const bool &enabled);
```
- `is_etag_enabled()` and `enable_etag(bool)` would be more idiomatic C++ naming
- `const bool &enabled` is unnecessary (plain `bool` is sufficient)

### 4. Excessive Nesting Depth
- The ETag processing logic (~80 lines) is embedded directly within `handle_file_request`
- This should be extracted into a separate private method for better readability and maintainability

### 5.  Code Duplication
```cpp
if (req. has_header("If-Match")) {
  header_value = req.get_header_value("If-Match");
} else if (req.has_header("If-None-Match")) {
  header_value = req.get_header_value("If-None-Match");
}
```
- This could be simplified using a ternary operator or a helper variable

---

## 🟢 Minor Issues

### 6. Test Code Structure
- The `for` loop in tests (`i = 0; i < 2`) is hard to read
- Consider using `TEST_P` (parameterized tests) for better clarity

### 7. Verbose Comments
